##### Importação de bibliotecas

In [201]:
import pandas as pd
import itertools

##### Importação das bases de dados que estão em CSV

In [202]:
municipio = pd.read_csv(filepath_or_buffer='./dados/d_municipio.csv')
tempo = pd.read_csv(filepath_or_buffer='./dados/d_tempo.csv')
origem = pd.read_csv(filepath_or_buffer='./dados/origem.csv')

##### Entendimento da base de dados

In [203]:
display(origem.head(1)) #27820
display(municipio.head(5))
display(tempo.head(1))

nu_notificacao  tp_entrada  tp_pop_liberdade  tp_pop_rua  tp_forma  \
0         2889325         1.0               2.0         2.0       1.0   

      dt_nascimento dt_diagnostico_sintoma  tp_situacao_encerramento  \
0  1999-04-04T00:00       2018-08-24T00:00                       2.0   

   co_uf_residencia_atual  co_municipio_residencia_atual  
0                    52.0                       522140.0

dmun_codibge dmun_uf_sigla dmun_uf_nome dmun_uf_nomex  \
0             1             -            -             -   
1        110000            RO     Rondônia      RONDONIA   
2        110001            RO     Rondônia      RONDONIA   
3        110002            RO     Rondônia      RONDONIA   
4        110003            RO     Rondônia      RONDONIA   

            dmun_municipio          dmun_municipiox        dmun_regiao_saude  \
0                        -                        -                        -   
1  Município ignorado - RO  MUNICIPIO IGNORADO - RO  Município ignorado - RO   
2    Alta Floresta D'Oeste    ALTA FLORESTA D'OESTE             Zona da Mata   
3                Ariquemes                ARIQUEMES           Vale do Jamari   
4                   Cabixi                   CABIXI                 Cone Sul   

        dmun_regiao_saudex         dmun_macro_regiao  \
0                        -                         -   
1  MUNICIPIO IGNORADO - RO   Município ignorado - RO   
2             ZONA DA MATA  Região não definida - RO   
3           VALE DO JAMARI  Região não definida - RO   
4                 CONE SUL  Região não definida - RO   

         dmun_macro_regiaox  ... dmun_fronteira_faixa_mun  \
0                         -  ...                        -   
1   MUNICIPIO IGNORADO - RO  ...                        N   
2  REGIAO NAO DEFINIDA - RO  ...                        S   
3  REGIAO NAO DEFINIDA - RO  ...                        N   
4  REGIAO NAO DEFINIDA - RO  ...                        S   

  dmun_fronteira_zona_mun dmun_capital_mun   dmun_regiao  dmun_regiaox  \
0                       -                -             -             -   
1                       N                N  Região Norte  REGIAO NORTE   
2                       S                N  Região Norte  REGIAO NORTE   
3                       N                N  Região Norte  REGIAO NORTE   
4                       S                N  Região Norte  REGIAO NORTE   

  dmun_pobreza_mun dmun_semiar_mun dmun_ride_regiao dmun_ride_regiaox  \
0                -               -                -                 -   
1                N               N                -                 -   
2                N               N                -                 -   
3                N               N                -                 -   
4                N               N                -                 -   

  dmun_pertence_ride  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  

[5 rows x 32 columns]

dte_id is_last_day_of_month year_semester  semester_number month_name  \
0  20100101                   no       2010-S1              1.0    Janeiro   

  month_abbreviation  week_in_month is_first_day_of_month  week_in_year  \
0                Jan            1.0                   yes           1.0   

  bimonthly_name  ...  date_short  year_month_number quarter_name  \
0             B1  ...  01/01/2010             201001           Q1   

                           date_full  year_month_abbreviation  day_in_month  \
0  Sexta-feira, 1 de Janeiro de 2010                 2010-Jan           1.0   

  day_abbreviation   year4             date_long year_of_week  
0              Sex  2010.0  1 de Janeiro de 2010       2010.0  

[1 rows x 31 columns]

In [204]:
display(origem.query(expr="co_municipio_residencia_atual == 522140").head(1))
display(municipio.query(expr="dmun_codibge == 522140").head(1))
display(tempo.query(expr="dte_id == 20180824").head(1))

nu_notificacao  tp_entrada  tp_pop_liberdade  tp_pop_rua  tp_forma  \
0         2889325         1.0               2.0         2.0       1.0   

      dt_nascimento dt_diagnostico_sintoma  tp_situacao_encerramento  \
0  1999-04-04T00:00       2018-08-24T00:00                       2.0   

   co_uf_residencia_atual  co_municipio_residencia_atual  
0                    52.0                       522140.0

dmun_codibge dmun_uf_sigla dmun_uf_nome dmun_uf_nomex dmun_municipio  \
5583        522140            GO        Goiás         GOIAS       Trindade   

     dmun_municipiox dmun_regiao_saude dmun_regiao_saudex dmun_macro_regiao  \
5583        TRINDADE           Central            CENTRAL      Centro Oeste   

     dmun_macro_regiaox  ... dmun_fronteira_faixa_mun dmun_fronteira_zona_mun  \
5583       CENTRO OESTE  ...                        N                       N   

     dmun_capital_mun          dmun_regiao         dmun_regiaox  \
5583                N  Região Centro-Oeste  REGIAO CENTRO-OESTE   

     dmun_pobreza_mun dmun_semiar_mun    dmun_ride_regiao   dmun_ride_regiaox  \
5583                N               N  Restante do Estado  RESTANTE DO ESTADO   

     dmun_pertence_ride  
5583                NaN  

[1 rows x 32 columns]

dte_id is_last_day_of_month year_semester  semester_number month_name  \
3157  20180824                   no       2018-S2              2.0     Agosto   

     month_abbreviation  week_in_month is_first_day_of_month  week_in_year  \
3157                Ago            4.0                    no          34.0   

     bimonthly_name  ...  date_short  year_month_number quarter_name  \
3157             B4  ...  24/08/2018             201808           Q3   

                              date_full  year_month_abbreviation  \
3157  Sexta-feira, 24 de Agosto de 2018                 2018-Ago   

      day_in_month day_abbreviation   year4             date_long year_of_week  
3157          24.0              Sex  2018.0  24 de Agosto de 2018       2018.0  

[1 rows x 31 columns]

##### Unificação das informações usando a função merge do Python

In [205]:
dbaMerge = pd.merge(left=origem,
                        right=municipio,
                        left_on='co_municipio_residencia_atual',
                        right_on='dmun_codibge',
                        how='left',
                        indicator=True)

##### Ajusta os dados para fazer um novo merge com a d_tempo

In [206]:
dbaMerge['dataFormatada'] = pd.to_datetime(dbaMerge['dt_diagnostico_sintoma']).dt.strftime('%Y%m%d').astype(int)

In [207]:
display(dbaMerge[dbaMerge['dataFormatada'].isna()])

Empty DataFrame
Columns: [nu_notificacao, tp_entrada, tp_pop_liberdade, tp_pop_rua, tp_forma, dt_nascimento, dt_diagnostico_sintoma, tp_situacao_encerramento, co_uf_residencia_atual, co_municipio_residencia_atual, dmun_codibge, dmun_uf_sigla, dmun_uf_nome, dmun_uf_nomex, dmun_municipio, dmun_municipiox, dmun_regiao_saude, dmun_regiao_saudex, dmun_macro_regiao, dmun_macro_regiaox, dmun_divisao_adm, dmun_divisao_admx, dmun_micro_regiao, dmun_micro_regiaox, dmun_regiao_metropolitana, dmun_regiao_metropolitanax, dmun_area_uf, dmun_area_mun, dmun_latitude_mun, dmun_longitude_mun, dmun_altitude_mun, dmun_amazonia_mun, dmun_fronteira_faixa_mun, dmun_fronteira_zona_mun, dmun_capital_mun, dmun_regiao, dmun_regiaox, dmun_pobreza_mun, dmun_semiar_mun, dmun_ride_regiao, dmun_ride_regiaox, dmun_pertence_ride, _merge, dataFormatada]
Index: []

[0 rows x 44 columns]

##### Filtrar somente o estado de Goiás para começar a parte analítica dos dados.
- Nesse arquivo, deve ser retornada a quantidade de casos novos de tuberculose pulmonar de forma mensal para cada município de Goiás.

In [208]:
dbaFull = pd.merge(left=dbaMerge,
                        right=tempo,
                        left_on='dataFormatada',
                        right_on='dte_id',
                        how='left').query(expr="dmun_uf_sigla in ['GO']")

In [209]:
display(dbaFull[dbaFull['month_name'].isna()])

nu_notificacao  tp_entrada  tp_pop_liberdade  tp_pop_rua  tp_forma  \
15              10046         1.0               NaN         NaN       1.0   
66                567         1.0               NaN         NaN       1.0   
71              61300         1.0               NaN         NaN       2.0   
170                28         1.0               NaN         NaN       1.0   
189             16053         1.0               2.0         2.0       1.0   
...               ...         ...               ...         ...       ...   
27642               1         1.0               2.0         2.0       2.0   
27682             111         1.0               NaN         NaN       1.0   
27699         3556784         1.0               2.0         2.0       1.0   
27710           60517         5.0               NaN         NaN       1.0   
27777           17163         3.0               NaN         NaN       1.0   

          dt_nascimento dt_diagnostico_sintoma  tp_situacao_encerramento  \
15     1975-05-31T00:00       2005-08-19T00:00                       1.0   
66     1973-05-29T00:00       2009-07-22T00:00                       1.0   
71     1953-04-06T00:00       2007-07-10T00:00                       1.0   
170    1943-03-03T00:00       2007-05-18T00:00                       1.0   
189    1938-04-12T00:00       2004-08-03T00:00                       1.0   
...                 ...                    ...                       ...   
27642  1935-08-20T00:00       2007-01-06T00:00                       1.0   
27682  1984-07-31T00:00       2001-12-27T00:00                       5.0   
27699  1997-02-20T00:00       1997-02-20T00:00                       5.0   
27710  1943-07-01T00:00       2007-04-10T00:00                       1.0   
27777  1978-03-10T00:00       2005-08-23T00:00                       1.0   

       co_uf_residencia_atual  co_municipio_residencia_atual  ...  date_short  \
15                       52.0                       520870.0  ...         NaN   
66                       52.0                       521880.0  ...         NaN   
71                       52.0                       522185.0  ...         NaN   
170                      52.0                       521020.0  ...         NaN   
189                      52.0                       520549.0  ...         NaN   
...                       ...                            ...  ...         ...   
27642                    52.0                       521310.0  ...         NaN   
27682                    52.0                       520396.0  ...         NaN   
27699                    52.0                       520840.0  ...         NaN   
27710                    52.0                       522185.0  ...         NaN   
27777                    52.0                       520140.0  ...         NaN   

      year_month_number quarter_name date_full year_month_abbreviation  \
15                  NaN          NaN       NaN                     NaN   
66                  NaN          NaN       NaN                     NaN   
71                  NaN          NaN       NaN                     NaN   
170                 NaN          NaN       NaN                     NaN   
189                 NaN          NaN       NaN                     NaN   
...                 ...          ...       ...                     ...   
27642               NaN          NaN       NaN                     NaN   
27682               NaN          NaN       NaN                     NaN   
27699               NaN          NaN       NaN                     NaN   
27710               NaN          NaN       NaN                     NaN   
27777               NaN          NaN       NaN                     NaN   

      day_in_month day_abbreviation year4 date_long year_of_week  
15             NaN              NaN   NaN       NaN          NaN  
66             NaN              NaN   NaN       NaN          NaN  
71             NaN              NaN   NaN       NaN          NaN  
170            NaN              NaN   N

In [210]:
dbaFull['month_name'] = dbaFull['month_name'].fillna('N/A').astype(str)

##### Aplicando condicionais mediante dicionário de dados informado no repositório.
- Uma coluna chamada município, na qual é esperado que contenha o nome de cada município por extenso.

In [211]:
# Tratativo tp_entrada
dbaFull['motivoEntrada'] = dbaFull.apply(lambda row:
                                         'Caso Novo' if row['tp_entrada'] == 1 else
                                         'Recidiva' if row['tp_entrada'] == 2 else
                                         'Reingresso após Abandono' if row['tp_entrada'] == 3 else
                                         'Não Sabe' if row['tp_entrada'] == 4 else
                                         'Transferência' if row['tp_entrada'] == 5 else
                                         'Pós-óbito' if row['tp_entrada'] == 6 else 'Não informado'
                                         , axis=1)

# Tratativa tp_pop_liberdade
dbaFull['popLiber'] = dbaFull.apply(lambda row:
                                         'Sim' if row['tp_pop_liberdade'] == 1 else
                                         'Não' if row['tp_pop_liberdade'] == 2 else
                                         'Ignorado' if row['tp_pop_liberdade'] == 9 else 'Não informado'
                                         , axis=1)

# Tratativa tp_pop_rua
dbaFull['popRua'] = dbaFull.apply(lambda row:
                                         'Sim' if row['tp_pop_rua'] == 1 else
                                         'Não' if row['tp_pop_rua'] == 2 else
                                         'Ignorado' if row['tp_pop_rua'] == 9 else 'Não informado'
                                         , axis=1)

# Tratativa tp_forma
dbaFull['formaTb'] = dbaFull.apply(lambda row:
                                         'Pulmonar' if row['tp_forma'] == 1 else
                                         'Extrapulmonar' if row['tp_forma'] == 2 else
                                         'Pulmonar + Extrapulmonar' if row['tp_forma'] == 3 else 'Não informado'
                                         , axis=1)

# Tratativa tp_situacao_encerramento
dbaFull['situaEnce'] = dbaFull.apply(lambda row:
                                         'Cura' if row['tp_situacao_encerramento'] == 1 else
                                         'Abandono' if row['tp_situacao_encerramento'] == 2 else
                                         'Óbito por TB' if row['tp_situacao_encerramento'] == 3 else
                                         'Óbito por outras causas' if row['tp_situacao_encerramento'] == 4 else
                                         'Transferência' if row['tp_situacao_encerramento'] == 5 else
                                         'Mudança de Diagnóstico' if row['tp_situacao_encerramento'] == 6 else
                                         'TB-DR' if row['tp_situacao_encerramento'] == 7 else
                                         'Mudança de Esquema' if row['tp_situacao_encerramento'] == 8 else 'Não informado'
                                         , axis=1)

dbaFull['ano'] = pd.to_datetime(dbaFull['dt_diagnostico_sintoma']).dt.strftime('%Y').astype(int)
dbaFull['mes'] = dbaFull.apply(lambda row: str(row['month_name']), axis=1).astype(str)

dbaFull['dtNascimento'] = pd.to_datetime(dbaFull['dt_nascimento']).dt.strftime('%Y-%m-%d')
dbaFull['dtDiagnostico'] = pd.to_datetime(dbaFull['dt_diagnostico_sintoma']).dt.strftime('%Y-%m-%d')

# dbaFull['ref'] = pd.to_datetime(dbaFull['dt_diagnostico_sintoma']).dt.strftime('%Y/%m')
# dbaFull['refExt'] = dbaFull.apply(lambda row: str(row['month_name']) + ' de ' + str(row['ano']), axis=1)

dbaFull['month_number'] = dbaFull['month_number'].fillna(0)

dbaFull.rename(columns={'dmun_municipio': 'municipio',
                        'dmun_uf_sigla': 'uf', 
                        'dmun_uf_nome': 'uf_nome',
                        'dmun_longitude_mun': 'longitude',
                        'dmun_latitude_mun': 'latitude'
                        }, inplace=True)

In [212]:
display(dbaFull.head(1))

nu_notificacao  tp_entrada  tp_pop_liberdade  tp_pop_rua  tp_forma  \
0         2889325         1.0               2.0         2.0       1.0   

      dt_nascimento dt_diagnostico_sintoma  tp_situacao_encerramento  \
0  1999-04-04T00:00       2018-08-24T00:00                       2.0   

   co_uf_residencia_atual  co_municipio_residencia_atual  ...  year_of_week  \
0                    52.0                       522140.0  ...        2018.0   

  motivoEntrada popLiber popRua   formaTb situaEnce   ano     mes  \
0     Caso Novo      Não    Não  Pulmonar  Abandono  2018  Agosto   

  dtNascimento dtDiagnostico  
0   1999-04-04    2018-08-24  

[1 rows x 84 columns]

In [213]:
print(dbaFull.columns)

Index(['nu_notificacao', 'tp_entrada', 'tp_pop_liberdade', 'tp_pop_rua',
       'tp_forma', 'dt_nascimento', 'dt_diagnostico_sintoma',
       'tp_situacao_encerramento', 'co_uf_residencia_atual',
       'co_municipio_residencia_atual', 'dmun_codibge', 'uf', 'uf_nome',
       'dmun_uf_nomex', 'municipio', 'dmun_municipiox', 'dmun_regiao_saude',
       'dmun_regiao_saudex', 'dmun_macro_regiao', 'dmun_macro_regiaox',
       'dmun_divisao_adm', 'dmun_divisao_admx', 'dmun_micro_regiao',
       'dmun_micro_regiaox', 'dmun_regiao_metropolitana',
       'dmun_regiao_metropolitanax', 'dmun_area_uf', 'dmun_area_mun',
       'latitude', 'longitude', 'dmun_altitude_mun', 'dmun_amazonia_mun',
       'dmun_fronteira_faixa_mun', 'dmun_fronteira_zona_mun',
       'dmun_capital_mun', 'dmun_regiao', 'dmun_regiaox', 'dmun_pobreza_mun',
       'dmun_semiar_mun', 'dmun_ride_regiao', 'dmun_ride_regiaox',
       'dmun_pertence_ride', '_merge', 'dataFormatada', 'dte_id',
       'is_last_day_of_month', 'year_s

##### Filtra apenas algumas colunas do dataset

In [214]:
filtro = ['nu_notificacao', 'tp_entrada', 'tp_pop_rua', 'tp_forma', 'tp_situacao_encerramento', 'dtNascimento', 'dtDiagnostico', 'motivoEntrada', 'popLiber', 'popRua', 'formaTb', 'situaEnce',
          'uf', 'uf_nome', 'municipio', 'latitude', 'longitude', 'ano', 'mes']

dbaFilter = dbaFull[filtro]

##### Aplica condicional:
- Apenas os pacientes que cuja o tipo de entrada em uma unidade seja classificado como "Caso novo", ou seja, todos os pacientes que nunca se submeteram ao tratamento de tuberculose. Ou então, aqueles pacientes cujo tipo de entrada é classificado como "Pós-óbito" ou cujo tipo de entrada é desconhecido.

- Cuja forma clínica de tuberculose é classificada como "Pulmonar" ou "Pulmonar + extrapulmonar".

- Podendo ser qualquer "Situação de encerramento" do caso notificado, exceto "Mudança de Diagnóstico".

- Apenas dados a partir de 2021.

In [215]:
condicional = dbaFilter.query("tp_entrada in [1, 4, 6] and tp_forma in [1, 3] and tp_situacao_encerramento not in [6] and ano > 2021")

In [216]:
display(dbaFilter, condicional)

nu_notificacao  tp_entrada  tp_pop_rua  tp_forma  \
0             2889325         1.0         2.0       1.0   
1                4922         1.0         NaN       1.0   
2             2763661         1.0         2.0       1.0   
3             2546548         1.0         2.0       1.0   
4             2873270         1.0         1.0       1.0   
...               ...         ...         ...       ...   
27815         2954272         1.0         2.0       3.0   
27816         2889306         1.0         2.0       3.0   
27817         3201314         1.0         2.0       1.0   
27818         2873263         1.0         2.0       1.0   
27819         2801389         1.0         2.0       2.0   

       tp_situacao_encerramento dtNascimento dtDiagnostico motivoEntrada  \
0                           2.0   1999-04-04    2018-08-24     Caso Novo   
1                           1.0   1981-04-02    2010-03-08     Caso Novo   
2                           1.0   1966-03-08    2017-09-18     Caso Novo   
3                           1.0   1970-10-04    2018-01-30     Caso Novo   
4                           1.0   1984-06-16    2018-05-24     Caso Novo   
...                         ...          ...           ...           ...   
27815                       1.0   1960-01-11    2017-08-30     Caso Novo   
27816                       1.0   1991-05-29    2018-06-03     Caso Novo   
27817                       1.0   1979-08-13    2018-08-06     Caso Novo   
27818                       5.0   1998-04-08    2018-06-21     Caso Novo   
27819                       2.0   1973-01-18    2018-01-17     Caso Novo   

            popLiber         popRua                   formaTb      situaEnce  \
0                Não            Não                  Pulmonar       Abandono   
1      Não informado  Não informado                  Pulmonar           Cura   
2                Não            Não                  Pulmonar           Cura   
3                Sim            Não                  Pulmonar           Cura   
4                Não            Sim                  Pulmonar           Cura   
...              ...            ...                       ...            ...   
27815            Não            Não  Pulmonar + Extrapulmonar           Cura   
27816            Não            Não  Pulmonar + Extrapulmonar           Cura   
27817            Não            Não                  Pulmonar           Cura   
27818            Não            Não                  Pulmonar  Transferência   
27819            Não            Não             Extrapulmonar       Abandono   

       uf uf_nome             municipio    latitude   longitude   ano  \
0      GO   Goiás              Trindade  -16.655350  -49.489100  2018   
1      GO   Goiás  Aparecida de Goiânia  -16.822150  -49.243330  2010   
2      GO   Goiás        Senador Canedo  -16.718580  -49.087970  2017   
3      GO   Goiás         Campos Verdes  -14.273270  -49.663270  2018   
4      GO   Goiás          Caldas Novas  -17.743540  -48.621200  2018   
...    ..     ...                   ...         ...         ...   ...   
27815  GO   Goiás               Goiânia  -16.673310  -49.255810  2017   
27816  GO   Goiás              Trindade  -16.655350  -49.489100  2018   
27817  GO   Goiás            Cavalcante  -13.799280  -47.463440  2018   
27818  GO   Goiás        Senador Canedo  -16.718580  -49.087970  2018   
27819  GO   Goiás           Doverlândia  -16.721040  -52.313980  2018   

            mes  
0        Agosto  
1         Março  
2      Setembro  
3       Janeiro  
4          Maio  
...         ...  
27815    Agosto  
27816     Junho  
27817    Agosto  
27818     Junho  
27819   Janeiro  

[26672 rows x 19 columns]

nu_notificacao  tp_entrada  tp_pop_rua  tp_forma  \
1990          5680141         1.0         2.0       1.0   
2795          5048484         1.0         2.0       1.0   
2796          5048485         1.0         2.0       1.0   
2847           245320         1.0         2.0       1.0   
3131          5198892         1.0         2.0       1.0   
...               ...         ...         ...       ...   
12044         5523737         1.0         2.0       1.0   
12403         5676558         1.0         2.0       1.0   
24077         5523757         1.0         2.0       1.0   
24252         5512014         1.0         2.0       3.0   
24678         4334084         1.0         2.0       1.0   

       tp_situacao_encerramento dtNascimento dtDiagnostico motivoEntrada  \
1990                        NaN   1968-11-03    2024-06-19     Caso Novo   
2795                        NaN   1957-04-06    2024-05-16     Caso Novo   
2796                        NaN   1973-04-29    2024-05-28     Caso Novo   
2847                        NaN   1964-01-23    2024-06-14     Caso Novo   
3131                        NaN   1964-05-18    2024-02-19     Caso Novo   
...                         ...          ...           ...           ...   
12044                       NaN   1977-05-23    2024-04-10     Caso Novo   
12403                       NaN   1971-06-24    2024-05-09     Caso Novo   
24077                       NaN   2005-07-05    2024-05-12     Caso Novo   
24252                       NaN   1980-09-28    2024-06-02     Caso Novo   
24678                       3.0   1950-03-09    2023-05-24     Caso Novo   

      popLiber popRua                   formaTb      situaEnce  uf uf_nome  \
1990       Não    Não                  Pulmonar  Não informado  GO   Goiás   
2795       Não    Não                  Pulmonar  Não informado  GO   Goiás   
2796       Não    Não                  Pulmonar  Não informado  GO   Goiás   
2847       Não    Não                  Pulmonar  Não informado  GO   Goiás   
3131       Não    Não                  Pulmonar  Não informado  GO   Goiás   
...        ...    ...                       ...            ...  ..     ...   
12044      Não    Não                  Pulmonar  Não informado  GO   Goiás   
12403      Não    Não                  Pulmonar  Não informado  GO   Goiás   
24077      Não    Não                  Pulmonar  Não informado  GO   Goiás   
24252      Não    Não  Pulmonar + Extrapulmonar  Não informado  GO   Goiás   
24678      Não    Não                  Pulmonar   Óbito por TB  GO   Goiás   

                   municipio    latitude   longitude   ano        mes  
1990   Águas Lindas de Goiás  -15.737940  -48.283440  2024      Junho  
2795               Rio Verde  -17.794540  -50.929680  2024       Maio  
2796               Rio Verde  -17.794540  -50.929680  2024       Maio  
2847            Quirinópolis  -18.451510  -50.452260  2024      Junho  
3131               Amaralina  -13.925480  -49.295380  2024  Fevereiro  
...                      ...         ...         ...   ...        ...  
12044                Goiânia  -16.673310  -49.255810  2024      Abril  
12403                 Itauçu  -16.203350  -49.601130  2024       Maio  
24077                Goiânia  -16.673310  -49.255810  2024       Maio  
24252    Bom Jardim de Goiás  -16.200690  -52.175370  2024      Junho  
24678                  Guapó  -16.831320  -49.536730  2023       Maio  

[2453 rows x 19 columns]

In [217]:
gbDf = condicional.groupby(['municipio', 'mes', 'ano']).size().reset_index(name='quantidade')
len(gbDf)

1069

##### Montando um dataframe auxiliar para tratar o ponto diferencial do projeto
- Será um diferencial caso consiga retornar para todos os períodos a partir de 2021, todos os 246 municípios de Goiás. Caso algum período para determinado município não tenha valores para serem apresentados, a coluna "Quantidade" deve retornar 0 (zero).

In [218]:
municipios = dbaFull['municipio'].unique()
mes = dbaFull['mes'].unique()
ano = dbaFull['ano'].unique()

print(len(municipios))

distinctMunicipio = pd.DataFrame(list(itertools.product(municipios, mes, ano)), columns=['municipio', 'mes', 'ano']).reset_index(drop=True)
display(distinctMunicipio)

247


municipio     mes   ano
0        Trindade  Agosto  2018
1        Trindade  Agosto  2010
2        Trindade  Agosto  2017
3        Trindade  Agosto  2012
4        Trindade  Agosto  2005
...           ...     ...   ...
147701  Goiandira   Abril  1979
147702  Goiandira   Abril  1950
147703  Goiandira   Abril  1965
147704  Goiandira   Abril  1960
147705  Goiandira   Abril  1997

[147706 rows x 3 columns]

In [219]:
display(gbDf)

municipio        mes   ano  quantidade
0           Abadia de Goiás     Agosto  2022           1
1           Abadia de Goiás  Fevereiro  2024           1
2           Abadia de Goiás       Maio  2023           1
3           Abadia de Goiás   Novembro  2023           1
4           Abadia de Goiás    Outubro  2023           1
...                     ...        ...   ...         ...
1064  Águas Lindas de Goiás      Março  2024           1
1065  Águas Lindas de Goiás   Novembro  2022           3
1066  Águas Lindas de Goiás   Novembro  2023           1
1067  Águas Lindas de Goiás    Outubro  2022           2
1068  Águas Lindas de Goiás   Setembro  2022           2

[1069 rows x 4 columns]

##### Aplicando left join para trazer todos os dados montados com os dois dataframes: o df com o condicional do projeto e o outro df com todos os municípios do estado de Goiás. Quem estiver com a quantidade nula, significa que para aquele período, o estado X não recebeu nenhuma notificação de paciente com Tuberculose. Nesse caso, o contador deve ser 0.

In [220]:
dfEnd = pd.merge(distinctMunicipio, gbDf, on=['municipio', 'mes', 'ano'], how='left').query(expr="ano > 2021")
dfEnd['quantidade'] = dfEnd['quantidade'].fillna(0).astype(int)
display(dfEnd)

municipio     mes   ano  quantidade
20       Trindade  Agosto  2024           0
24       Trindade  Agosto  2023           3
26       Trindade  Agosto  2022           4
66       Trindade   Março  2024           1
70       Trindade   Março  2023           3
...           ...     ...   ...         ...
147638  Goiandira   Julho  2023           0
147640  Goiandira   Julho  2022           0
147680  Goiandira   Abril  2024           0
147684  Goiandira   Abril  2023           0
147686  Goiandira   Abril  2022           0

[9633 rows x 4 columns]

##### Validando a quantidade de estados que está no dataframe. Tem que ser igual a 247

In [221]:
distinctMunicipioTeste = dfEnd['municipio'].drop_duplicates()

display(distinctMunicipioTeste)

20                      Trindade
618         Aparecida de Goiânia
1216              Senador Canedo
1814               Campos Verdes
2412                Caldas Novas
                   ...          
144736         Ipiranga de Goiás
145334    Santa Rita do Araguaia
145932                  Guaraíta
146530              Amorinópolis
147128                 Goiandira
Name: municipio, Length: 247, dtype: object

##### Criando o campo período concatenando o mes e o ano da notificação.
- Uma coluna chamada mês com o período por extenso no seguinte formato: "<Mês> de <Ano>", exemplo: "Janeiro de 2024".

In [222]:
dfEnd['periodo'] = dfEnd['mes'].astype(str) + ' de ' + dfEnd['ano'].astype(str)

In [223]:
display(dfEnd['mes'].drop_duplicates())

20        Agosto
66         Março
112     Setembro
158      Janeiro
204         Maio
250        Junho
296    Fevereiro
342      Outubro
388          N/A
434     Novembro
480     Dezembro
526        Julho
572        Abril
Name: mes, dtype: object

In [224]:
dfEnd = dfEnd.sort_values(by=['ano', 'mes', 'municipio', 'quantidade'], ascending=True)
dfEnd = dfEnd[['municipio', 'periodo', 'quantidade']]
display(dfEnd)

municipio           periodo  quantidade
44830         Abadia de Goiás     Abril de 2022           0
38850               Abadiânia     Abril de 2022           3
13136                 Acreúna     Abril de 2022           1
96258               Adelândia     Abril de 2022           0
76524                Alexânia     Abril de 2022           0
...                       ...               ...         ...
49746                Vila Boa  Setembro de 2024           0
89812           Vila Propício  Setembro de 2024           0
122702     Água Fria de Goiás  Setembro de 2024           0
141838             Água Limpa  Setembro de 2024           0
9082    Águas Lindas de Goiás  Setembro de 2024           0

[9633 rows x 3 columns]

##### Eliminando período onde o mes é N/A (Não existe)

In [225]:
dfEnd = dfEnd[~dfEnd['periodo'].str.contains("N/A")]

##### Última validação da quantidade de monicípios no dataframe final (dfEnd)

In [226]:
display(len(municipio.query(expr="dmun_uf_sigla in ['GO']")['dmun_municipio'].drop_duplicates()))

display(len(distinctMunicipio['municipio'].drop_duplicates()))

display(len(dfEnd['municipio'].drop_duplicates()))

247

247

247

In [227]:
display(len(dfEnd['periodo'].drop_duplicates()))

36

In [228]:
# Total de linhas - Distict período X Distinct município

display(len(dfEnd['municipio'].drop_duplicates()) * len(dfEnd['periodo'].drop_duplicates()))

# Total de linha do dataframe final
display(len(dfEnd))

## Se os valores baterem, é possível concluir que existe uma contagem de casos de Tuberculose para todos os municípios em todos os períodos que houve notificação.
## Se não houver nenhum caso para um município específico, ele será demonstrado como 0

8892

8892

##### Gera a base final e uma base para validação

In [229]:
dbaFilter.to_csv(path_or_buf="./output/validacao.csv", index=False)

In [230]:
dfEnd.to_csv(path_or_buf="./output/saida.csv", index=False)

##### Identifica alguns casos específicos

In [231]:
display(len(dbaFilter.query(expr="motivoEntrada in ['Não informado']")))

5

- 5 pacientes com o campo tp_entrada não preenchidos. Esse campo é considerado obrigadtório.
Nesse caso, essa quantidade entraria na contagem de desconhecidos ou o correto seria manter apenas os registros com motivoEntrada "Não Sabe" como desconhecido?

In [232]:
display(len(dbaFilter.query(expr="formaTb in ['Não informado']")))

5

In [233]:
display(dbaFilter.query(expr="motivoEntrada in ['Não informado']"))

nu_notificacao  tp_entrada  tp_pop_rua  tp_forma  \
10687           48051         NaN         NaN       NaN   
11916             245         NaN         NaN       NaN   
12344             923         NaN         NaN       NaN   
23500           15029         NaN         NaN       NaN   
25468            1307         NaN         NaN       NaN   

       tp_situacao_encerramento dtNascimento dtDiagnostico  motivoEntrada  \
10687                       1.0   1967-11-17    2006-02-24  Não informado   
11916                       5.0   1980-07-11    2005-09-27  Não informado   
12344                       1.0   1961-08-17    2006-01-30  Não informado   
23500                       1.0   1953-04-02    2003-03-22  Não informado   
25468                       1.0   1938-02-16    2005-11-21  Não informado   

            popLiber         popRua        formaTb      situaEnce  uf uf_nome  \
10687  Não informado  Não informado  Não informado           Cura  GO   Goiás   
11916  Não informado  Não informado  Não informado  Transferência  GO   Goiás   
12344  Não informado  Não informado  Não informado           Cura  GO   Goiás   
23500  Não informado  Não informado  Não informado           Cura  GO   Goiás   
25468  Não informado  Não informado  Não informado           Cura  GO   Goiás   

                   municipio    latitude   longitude   ano  mes  
10687            Castelândia  -18.089850  -50.229250  2006  N/A  
11916               Itaberaí  -16.021130  -49.803480  2005  N/A  
12344         Cachoeira Alta  -18.751000  -50.944710  2006  N/A  
23500  Campo Alegre de Goiás  -17.640560  -47.778230  2003  N/A  
25468              Itumbiara  -18.417680  -49.220870  2005  N/A

- Esses 5 pacientes também não possui a tp_forma preenchidos, tornando inviável saber a forma clínica da Tuberculose dos mesmos. Nesse sentido, acredito que o correto seria remover eles da contagem, como o script ja está fazendo. Ate esse momento, a validação das bases está seguindo o fluxo que deveria seguir.